File: php/DataStructureBenchmark/benchmark.php
<?php

declare(strict_types=1);

require_once __DIR__ . '/structures/BTree.php';
require_once __DIR__ . '/structures/HashTable.php';
require_once __DIR__ . '/structures/LinkedList.php';
require_once __DIR__ . '/structures/ArrayList.php';
require_once __DIR__ . '/structures/Vector.php';
require_once __DIR__ . '/structures/Graph.php';
require_once __DIR__ . '/structures/Heap.php';

const DATA_FILE = __DIR__ . '/data/dataset.txt';
const RESULT_DIR = __DIR__ . '/results';
const RESULT_FILE = RESULT_DIR . '/benchmark_result.txt';

// 검색 샘플 개수
const SEARCH_SAMPLE_SIZE = 1000;

// 정렬 실행 여부
const ENABLE_SORT = true;

function ensureDirectory(string $dir): void
{
    if (!is_dir($dir)) {
        mkdir($dir, 0777, true);
    }
}

function nowMs(): float
{
    return microtime(true) * 1000;
}

/**
 * 검색용 key (UUID) 샘플 로드
 */
function loadSearchKeys(string $file, int $sampleSize = SEARCH_SAMPLE_SIZE): array
{
    $keys = [];
    $fp = fopen($file, 'rb');

    if ($fp === false) {
        throw new RuntimeException('데이터 파일을 열 수 없습니다.');
    }

    // 헤더 스킵
    fgets($fp);

    while (($line = fgets($fp)) !== false && count($keys) < $sampleSize) {
        $parts = str_getcsv(trim($line), ',', '"', '\\');

        if (count($parts) >= 1) {
            $keys[] = $parts[0]; // UUID
        }
    }

    fclose($fp);
    return $keys;
}

/**
 * 각 자료구조 벤치마크 실행
 */
function benchmarkStructure(string $name, object $structure, string $dataFile, array $searchKeys): string
{
    // ------------------
    // INSERT
    // ------------------
    $insertStart = nowMs();

    $fp = fopen($dataFile, 'rb');

    if ($fp === false) {
        throw new RuntimeException('데이터 파일을 열 수 없습니다.');
    }

    // 헤더 스킵
    fgets($fp);

    $count = 0;

    while (($line = fgets($fp)) !== false) {

        $parts = str_getcsv(trim($line), ',', '"', '\\');

        // ⭐ 3컬럼 처리
        if (count($parts) !== 3) {
            continue;
        }

        [$key, $keyString, $value] = $parts;

        $structure->insert($key, $keyString, $value);

        $count++;
    }

    fclose($fp);

    $insertEnd = nowMs();
    $insertTime = $insertEnd - $insertStart;

    // ------------------
    // SEARCH
    // ------------------
    $searchStart = nowMs();

    $found = 0;

    foreach ($searchKeys as $key) {
        $result = $structure->search($key);

        if ($result !== null) {
            $found++;
        }
    }

    $searchEnd = nowMs();
    $searchTime = $searchEnd - $searchStart;

    // ------------------
    // SORT
    // ------------------
    $sortTime = 0.0;

    if (ENABLE_SORT && method_exists($structure, 'sortByKeyString') && $name !== 'BTree') {
        $sortStart = nowMs();

        $structure->sortByKeyString();

        $sortEnd = nowMs();
        $sortTime = $sortEnd - $sortStart;
    }

    // ------------------
    // MEMORY
    // ------------------
    $peakMemory = memory_get_peak_usage(true) / 1024 / 1024;

    // ------------------
    // RESULT
    // ------------------
    $result = [];
    $result[] = "========== {$name} ==========";
    $result[] = "Inserted Count     : {$count}";
    $result[] = "Insert Time (ms)   : " . number_format($insertTime, 3);
    $result[] = "Search Sample      : " . count($searchKeys);
    $result[] = "Found Count        : {$found}";
    $result[] = "Search Time (ms)   : " . number_format($searchTime, 3);
    $result[] = "Sort Time (ms)     : " . number_format($sortTime, 3);
    $result[] = "Peak Memory (MB)   : " . number_format($peakMemory, 3);
    $result[] = "";

    return implode(PHP_EOL, $result);
}

// ------------------
// 실행
// ------------------

if (!file_exists(DATA_FILE)) {
    exit("data/dataset.txt 가 없습니다. 먼저 generate_data.php를 실행하세요." . PHP_EOL);
}

ensureDirectory(RESULT_DIR);

// 검색용 key 샘플
$searchKeys = loadSearchKeys(DATA_FILE);

// 자료구조 목록
$structures = [
    'HashTable'  => new HashTable(),
    'BTree'      => new BTree(8),
    'LinkedList' => new LinkedList(),
    'ArrayList'  => new ArrayList(),
    'Vector'     => new Vector(),
    'Graph'      => new Graph(),
    'Heap'       => new HeapStructure(),
];

$output = [];
$output[] = "Benchmark Started At: " . date('Y-m-d H:i:s');
$output[] = "Data File: " . DATA_FILE;
$output[] = "Search Sample Size: " . count($searchKeys);
$output[] = "";

foreach ($structures as $name => $structure) {

    gc_collect_cycles();

    $output[] = benchmarkStructure($name, $structure, DATA_FILE, $searchKeys);

    unset($structure);

    gc_collect_cycles();
}

file_put_contents(RESULT_FILE, implode(PHP_EOL, $output));

echo "벤치마크 완료: " . RESULT_FILE . PHP_EOL;

File: java/DataStructureBenchmark/src/main/Benchmark.java
package main;


import java.io.BufferedReader;
import java.io.File;
import java.io.FileReader;
import java.io.FileWriter;
import java.io.IOException;
import java.util.ArrayList;
import java.util.Date;
import java.util.LinkedHashMap;
import java.util.List;
import java.util.Map;
import java.util.UUID;
import java.lang.management.ManagementFactory;
import java.lang.management.MemoryMXBean;
import java.lang.management.MemoryUsage;

import structures.Arraylist.Arraylist;
import structures.BTree.BTree;
import structures.Graph.Graph;
import structures.HashTable.HashTable;
import structures.Heap.Heap;
import structures.LinkedList.LinkedList;
import structures.Vector.Vector;

public class Benchmark {
    public static void main(String[] args) throws IOException {
        String rootDir = "C:/Users/jeonghwanroh/algorithmTest/java/DataStructureBenchmark";
        String dataFile = rootDir + "/data/dataset.txt";
        String resultFile = rootDir + "/results/benchmark_result.txt";
        int searchSampleSize = 1000;

        // 자료구조 인스턴스 생성 (클래스는 앞서 제공한 Java 버전 사용)
        HashTable hashTable = new HashTable();
        BTree bTree = new BTree(8);
        LinkedList linkedList = new LinkedList();
        Arraylist arrayList = new Arraylist();
        Vector vector = new Vector();
        Graph graph = new Graph();
        Heap heap = new Heap();

        // Generate data dynamically
        int dataCount = 200000; // 20 million entries
        List<String[]> data = generateData(dataCount);

        // Ensure the 'data' directory exists
        File dataDir = new File(rootDir + "/data");
        if (!dataDir.exists()) {
            dataDir.mkdirs();
        }

        // Save generated data to a file
        try (FileWriter fw = new FileWriter(dataFile)) {
            fw.write("key,keyString,value\n"); // Header
            for (String[] row : data) {
                fw.write(String.join(",", row) + "\n");
            }
        }
        System.out.println("Data generation completed: " + dataFile);

        // 샘플 검색 키 추출
        List<String> searchKeys = new ArrayList<>();
        for (int i = 0; i < Math.min(searchSampleSize, data.size()); i++) {
            searchKeys.add(data.get(i)[0]);
        }

        // 벤치마크 실행
        Map<String, Object> structures = new LinkedHashMap<>();
        structures.put("HashTable", hashTable);
        structures.put("BTree", bTree);
        structures.put("LinkedList", linkedList);
        structures.put("ArrayList", arrayList);
        structures.put("Vector", vector);
        structures.put("Graph", graph);
        structures.put("Heap", heap);

        // Ensure the 'results' directory exists
        File resultsDir = new File(rootDir + "/results");
        if (!resultsDir.exists()) {
            resultsDir.mkdirs();
        }

        MemoryMXBean memoryBean = ManagementFactory.getMemoryMXBean();

        StringBuilder output = new StringBuilder();
        output.append("Benchmark Started At: ").append(new Date()).append("\n");
        output.append("Data File: ").append(dataFile).append("\n");
        output.append("Search Sample Size: ").append(searchKeys.size()).append("\n\n");

        for (Map.Entry<String, Object> entry : structures.entrySet()) {
            String name = entry.getKey();
            Object structure = entry.getValue();
            try {
                long insertStart = System.currentTimeMillis();
                for (String[] row : data) {
                    if (structure instanceof HashTable)
                        ((HashTable) structure).insert(row[0], row[1], row[2]);
                    else if (structure instanceof BTree)
                        ((BTree) structure).insert(row[0], row[1], row[2]);
                    else if (structure instanceof LinkedList)
                        ((LinkedList) structure).insert(row[0], row[1], row[2]);
                    else if (structure instanceof Arraylist)
                        ((Arraylist) structure).insert(row[0], row[1], row[2]);
                    else if (structure instanceof Vector)
                        ((Vector) structure).insert(row[0], row[1], row[2]);
                    else if (structure instanceof Graph)
                        ((Graph) structure).insert(row[0], row[1], row[2]);
                    else if (structure instanceof Heap)
                        ((Heap) structure).insert(row[0], row[1], row[2]);
                }
                long insertEnd = System.currentTimeMillis();
                long insertTime = insertEnd - insertStart;

                MemoryUsage heapMemoryUsage = memoryBean.getHeapMemoryUsage();
                long peakMemory = heapMemoryUsage.getUsed() / (1024 * 1024);

                long sortStart = System.currentTimeMillis();
                if (structure instanceof Arraylist)
                    ((Arraylist) structure).sort();
                else if (structure instanceof Vector)
                    ((Vector) structure).sort();
                else if (structure instanceof LinkedList)
                    ((LinkedList) structure).sort();
                long sortEnd = System.currentTimeMillis();
                long sortTime = sortEnd - sortStart;

                long searchStart = System.currentTimeMillis();
                int found = 0;
                for (String key : searchKeys) {
                    Object result = null;
                    if (structure instanceof HashTable)
                        result = ((HashTable) structure).search(key);
                    else if (structure instanceof BTree)
                        result = ((BTree) structure).search(key);
                    else if (structure instanceof LinkedList)
                        result = ((LinkedList) structure).search(key);
                    else if (structure instanceof Arraylist)
                        result = ((Arraylist) structure).search(key);
                    else if (structure instanceof Vector)
                        result = ((Vector) structure).search(key);
                    else if (structure instanceof Graph)
                        result = ((Graph) structure).search(key);
                    else if (structure instanceof Heap)
                        result = ((Heap) structure).search(key);
                    if (result != null) found++;
                }
                long searchEnd = System.currentTimeMillis();
                long searchTime = searchEnd - searchStart;

                output.append("========== ").append(name).append(" ==========").append("\n");
                output.append("Inserted Count     : ").append(data.size()).append("\n");
                output.append("Insert Time (ms)   : ").append(insertTime).append("\n");
                output.append("Search Sample      : ").append(searchKeys.size()).append("\n");
                output.append("Found Count        : ").append(found).append("\n");
                output.append("Search Time (ms)   : ").append(searchTime).append("\n");
                output.append("Sort Time (ms)     : ").append(sortTime).append("\n");
                output.append("Peak Memory (MB)   : ").append(peakMemory).append("\n\n");

                // Write partial results to file
                try (FileWriter fw = new FileWriter(resultFile, true)) {
                    fw.write(output.toString());
                }
                output.setLength(0); // Clear the buffer

                System.out.println(name + " benchmark completed.");
            } catch (Exception e) {
                System.err.println("Error during " + name + " benchmark: " + e.getMessage());
            }
        }
    }

    private static List<String[]> generateData(int count) {
        List<String[]> data = new ArrayList<>();
        for (int i = 0; i < count; i++) {
            String key = UUID.randomUUID().toString();
            String keyString = "KeyString_" + i;
            String value = "Value_" + UUID.randomUUID().toString();
            data.add(new String[]{key, keyString, value});
        }
        return data;
    }
}